### PR TITLE
Audit Log to CloudWatch

### PR DESCRIPTION
**Pull request by joehoyle**
Initially this was part of the security module, but I think it actually makes a lot more sense here. This switches out the Stream DB provider for AWS CLoudWatch Logs in the cloud environments.

### CODE REVIEW

File: inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php [hunk 1]
```diff
@@ -0,0 +1,216 @@
+<?php
```
Comment (roborourke) [inline inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php]: missing file header comment

File: inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+		$results = [ 'status' => 'Running' ];
+		while ( ! in_array( $results['status'], [ 'Failed', 'Cancelled', 'Complete' ], true ) ) {
+			$results = cloudwatch_logs_client()->getQueryResults([
+				'queryId' => $query['queryId'],
+			] );
+		}
+
+		/**
+		 * Future optimization: currently we have to pull all results up until the offset.
+		 * There's no way in the CloudWatch Insights API to supply a number offset.
```
[the remark this answers is not in the document]
Comment (joehoyle) [reply inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php]: Yeah I'm not totally sure, I thought at least on the plus side, all the data _gets_ to Logs without a problem, it's a retrieval just an admin retrieval problem, so we could get the data if needed for further back. Perhaps I could test this with more scale though, to see.

File: inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+			'context',
+			'action',
+			'ip',
+			'object_id',
+		];
+
+		$field_where = [];
+		foreach ( $fields as $field ) {
+			if ( $args[ $field ] || $args[ $field ] === '0' ) {
+				$field_where[] = sprintf( '%s = "%s" ', $field, $args[ $field ] );
```
[the remark this answers is not in the document]
Comment (joehoyle) [reply inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php]: Probably not back to add escaping either way, will add.

File: inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (joehoyle) [reply inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php]: Reached out to AWS to ask about perhaps a more performant way to do this.

File: inc/audit_log_to_cloudwatch/namespace.php [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * Set the Stream DB Driver to the CloudWatch integration.
+ *
+ * This is only done when we we are deemed to have a connection to CloudWatch available.
+ *
+ * @param string $db_driver
+ * @return string
+ */
+function set_stream_db_driver( string $db_driver ) : string {
```
[the remark this answers is not in the document]
Comment (joehoyle) [reply inc/audit_log_to_cloudwatch/namespace.php]: I think it's better to be explicit with it

File: inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php (hunk 1 above)
Comment (joehoyle) [reply inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php]: We don't really have these elsewhere. Is this a requirement? Examples in our coding guidelines maybe?

**Comment by joehoyle** [comment]
Ok, waiting on hearing back from AWS to see about the 10k limit of retrieving entries we have.

File: inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php (hunk 1 above)
Comment (roborourke) [reply inc/audit_log_to_cloudwatch/class-cloudwatch-driver.php]: Oh, genuinely thought it was part of our coding standards. I've been doing it out of habit, a brief description and a `@package hm-platform/<module>` line.

**Comment by roborourke** [comment]
@joehoyle good to go for now if you want to pick up the offset problem with another issue?

**Comment by joehoyle** [comment]
It looks like it could be a show-stopper, so just waiting to see from AWS what the best path forward is there

**Comment by joehoyle** [comment]
I think this is actually possible if we push an "auto increment" value on each push to the log, this will give us a numeric ID in log objects, which means we can do the offset based off a `WHERE` clause in the query. For example `auto_increment > 10000 and auto_increment < 10010` would give you an offset.

We already have to store the sequencetoken for the cloudwatch stream, so we should be able to store the auto_increment "along side" it.

**Comment by joehoyle** [comment]
Hmm actually no, an auto_increment value won't work as there's other where conditions that make that useless. I also heard back from AWS who said indeed there is no way to paginate the results, or get more than 10k entries at once.

**Comment by joehoyle** [comment]
Ok I updated this to max out at 10k entries, which is a lot simpler. The user just needs to narrow down their search to a timeframe, or say specific user if they need to see beyond 10k items.

**Pull request merged**